src/ruby/puppetserver-lib/puppet/server/settings.rb

```diff
@@ -0,0 +1,85 @@
+require 'forwardable'
+
+class Puppet::Server::CompilerContext
+  THREAD_LOCAL_SETTINGS = Set.new([:strict, :code, :tasks])
+
+  def initialize(global_settings)
+    @settings = {}
+    THREAD_LOCAL_SETTINGS.each do |setting|
+      @settings[setting] = global_settings[setting]
+    end
+  end
+
+  def [](key)
+    @settings[key]
+  end
+end
+
+class Puppet::Server::Settings
+  extend Forwardable
+  THREAD_LOCAL_SETTINGS = Set.new([:strict, :code, :tasks])
+
+  def initialize(global_settings, puppet_config)
+    @global_settings = global_settings
+    @local_settings = Puppet::ThreadLocal.new {nil}
+    @puppet_config = puppet_config.reduce([]) do |acc, entry|
+      acc << "--#{entry[0]}"
+      acc << entry[1] if entry[1].kind_of?(String)
+      acc
+    end
+    @compiler_context = Puppet::ThreadLocal.new(Puppet::Server::CompilerContext.new(global_settings))
+  end
+
+  def_delegators :current_settings, :preferred_run_mode, :clear_environment_settings, :values, :parse_file,
```

Comment:
These delegated methods are ostensibly the "public" API of Puppet::Settings that are needed for basic compilations to work. I initially thought it would be a smaller number of delegated methods, but it seems to have grown to be a bit larger than I anticipated. Any thoughts on just inheriting from the Puppet::Settings rather than delegating all these methods?

Reply:
Just moved to the stand method_missing approach to delegate the to the correct settings object.